racket/collects/racket/hash.rkt

```diff
@@ -64,15 +66,15 @@
         res)))
 
 (define (hash-filter ht pred)
-  (cond [(immutable? ht)
-         (for ([(k v) (in-hash ht)])
-           (when (not (pred k v))
-             (set! ht (hash-remove ht k))))]
-        [else
-         (for ([(k v) (in-hash ht)])
-           (when (not (pred k v))
-             (hash-remove! ht k)))])
-  ht)
+  (define fresh-ht (hash-copy-clear ht))
```

Comment:
I think the change that @AlexKnauth had in mind is:

(*Caution: code not tested*)

```
(define (hash-filter ht pred)
  (cond
    [(immutable? ht)
     (for/fold ([ht ht]) ([(k v) (in-immutable-hash ht)])
       (cond
         [(pred k v) ht]
         [else (hash-remove ht k)]))]
    [else
     (define new-ht (hash-copy-clear ht))
     (for ([(k v) (in-hash ht)] #:when (pred k v))
       (hash-set! new-ht k v))
     new-ht]))
```

In particular:
- Avoid `set!`. Use `for/fold` to communicate to the next iteration, rather than via mutation.
- `in-immutable-hash` must appear under `for` syntactically for the optimization to kick in, so you can't pass it as an argument like that.
- Definitely avoid stuff from `racket/function`, which somewhat increases the start up time (IIRC).

Reply:
Thanks, will update later tonight or tomorrow morning. Lots to learn! I appreciate you two giving all this feedback

Reply:
If `#:when` works for one, it should work for both. Or maybe `#:unless`, I'm not sure.